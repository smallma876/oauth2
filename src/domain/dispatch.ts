export interface DispatchObject<T, P = any> {
  type: T;
  payload?: P;
}
