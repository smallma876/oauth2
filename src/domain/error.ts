export interface OauthError {
  message: string;
  status: number;
}
