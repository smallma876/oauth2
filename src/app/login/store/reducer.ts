import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { DispatchObject } from '../../../domain/dispatch';

export type DispatchLogin = Dispatch<DispatchObject<LoginActions>>;

export enum LoginActions {
  ClearData = 'CLEAR_DATA',
}

export interface LoginState {}

export const loginState: LoginState = {};

export const loginReducer = produce((draft: Draft<LoginState>, { type, payload }: DispatchObject<LoginActions>) => {
  switch (type) {
    case LoginActions.ClearData:
      break;
    default:
      throw new Error(`Unknown type: ${type}`);
  }
});
