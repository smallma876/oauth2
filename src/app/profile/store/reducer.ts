import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { DispatchObject } from '../../../domain/dispatch';

export type DispatchProfile = Dispatch<DispatchObject<ProfileActions>>;

export enum ProfileActions {
  ClearData = 'CLEAR_DATA',
}

export interface ProfileState {}

export const profileState: ProfileState = {};

export const profileReducer = produce(
  (draft: Draft<ProfileState>, { type, payload }: DispatchObject<ProfileActions>) => {
    switch (type) {
      case ProfileActions.ClearData:
        break;
      default:
        throw new Error(`Unknown type: ${type}`);
    }
  },
);
