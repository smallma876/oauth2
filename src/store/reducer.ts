import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { LoginToken } from '../app/login/domain/login-token';
import { DispatchObject } from '../domain/dispatch';

export type DispatchApp = Dispatch<DispatchObject<AppActions>>;

export enum AppActions {
  IsFetching = 'IS_FETCHING',
  FinishFetching = 'FINISHED_FETCHING',
  ErrorFetching = 'ERROR_FETCHING',
  SessionExpired = 'SESSION_EXPIRED',
  ShowLoadingBeforeNavigating = 'SHOW_LOADING_BEFORE_NAVIGATING',
  ClearError = 'CLEAR_ERROR',
  ClearData = 'CLEAR_DATA',
  SetUser = 'SET_USER',
  SetToken = 'SET_TOKEN',
}

export interface AppState {
  isFetching: boolean;
  pendingFetches: number;
  sessionExpired: boolean;
  showLoadingBeforeNavigating: boolean;
  user: LoginToken | null;
  token: string;
}

export const initialState: AppState = {
  isFetching: false,
  pendingFetches: 0,
  sessionExpired: false,
  showLoadingBeforeNavigating: false,
  user: null,
  token: '',
};

export const appReducer = produce((draft: Draft<AppState>, { type, payload }: DispatchObject<AppActions>) => {
  switch (type) {
    case AppActions.IsFetching:
      draft.pendingFetches += 1;
      draft.isFetching = true;
      break;
    case AppActions.FinishFetching:
      draft.pendingFetches -= 1;
      draft.isFetching = false;
      break;
    case AppActions.SessionExpired:
      draft.sessionExpired = true;
      break;
    case AppActions.SetUser:
      draft.user = payload as LoginToken;
      break;
    case AppActions.SetToken:
      draft.token = payload as string;
      break;
    default:
      throw new Error(`Unknown type: ${type}`);
  }
});
