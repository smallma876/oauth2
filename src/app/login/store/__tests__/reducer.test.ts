import { OauthError } from '../../../../domain/error';
import { LoginActions, loginReducer, loginState } from '../reducer';

describe('login reducer test', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should have fetching true', () => {
    const state = { ...loginState };
    const finalState = loginReducer(state, { type: LoginActions.ClearData });

    expect(finalState).toMatchSnapshot();
    expect(finalState).toEqual(state);
  });

  it('should return error', () => {
    try {
      const state = { ...loginState };
      loginReducer(state, { type: 'test' as LoginActions });
    } catch (error) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect((error as OauthError).message).toBe(`Unknown type: test`);
    }
  });
});
