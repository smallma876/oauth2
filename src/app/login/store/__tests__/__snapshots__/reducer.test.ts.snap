// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login reducer test should have fetching true 1`] = `Object {}`;
